relay-metrics/src/aggregator/inner.rs

```diff
@@ -0,0 +1,926 @@
+        );
+
+        let key_cost = key.cost() as u64;
+        match slot.buckets.entry(key) {
+            Entry::Occupied(occupied_entry) => {
+                let estimated_cost = match &value.value {
+                    // Counters and Gauges aggregate without additional costs.
+                    BucketValue::Counter(_) | BucketValue::Gauge(_) => 0,
+                    // Distributions are an accurate estimation, all values will be added.
+                    BucketValue::Distribution(d) => d.len() * mem::size_of::<DistributionType>(),
+                    // Sets are an upper bound.
+                    BucketValue::Set(s) => s.len() * mem::size_of::<SetType>(),
+                };
+
+                // Reserve for the upper bound of the value.
+                let reservation = slot.stats.reserve(
+                    &mut self.stats,
+                    project_key,
+                    namespace,
+                    estimated_cost as u64,
+                    &self.limits,
+                )?;
+
+                let actual_cost = occupied_entry.into_mut().merge(value)?;
+
+                // Track the actual cost increase, not just the reservation.
+                reservation.consume_with(actual_cost as u64);
+                slot.stats.incr_merges(namespace);
+            }
+            Entry::Vacant(vacant_entry) => {
+                let reservation = slot.stats.reserve(
+                    &mut self.stats,
+                    project_key,
+                    namespace,
+                    key_cost + value.value.cost() as u64,
+                    &self.limits,
+                )?;
+
+                vacant_entry.insert(value);
+
+                reservation.consume();
+                slot.stats.incr_count(&mut self.stats, namespace);
+            }
+        };
+
+        debug_assert_eq!(slot.stats.count, slot.buckets.len() as u64);
+
+        Ok(())
+    }
+
+    /// FLushes the next partition and advances time.
+    pub fn flush_next(&mut self) -> Partition {
+        let mut slot @ Slot { partition_key, .. } = self
+            .slots
+            .pop_front()
+            .expect("there should always be at least one partition");
+
+        let stats = PartitionStats::from(&slot.stats);
+
+        // Remove the partition cost from the total cost and reset the partition cost.
+        self.stats.remove_slot(&slot.stats);
+        slot.stats.reset();
+
+        // Create a new and empty slot with a similar capacity as the original one,
+        // but still shrinking a little bit so that over time we free up memory again
+        // without re-allocating too many times.
+        //
+        // Also re-use the original hasher, no need to create a new/expensive one.
+        self.slots.push_back(Slot {
+            buckets: HashMap::with_capacity_and_hasher(
+                slot.buckets.len(),
+                slot.buckets.hasher().clone(),
+            ),
+            ..slot
+        });
```

Comment:
I wonder if this would be simpler with an actual statically sized ring buffer. Then you could reset the slot at `head` and move the head, without `pop / push`.

Reply:
I also get the impression that this implementation has features of a queue and a static ring buffer simultaneously.

Reply:
You can rotate the `VecDeque`, initially this is what I had, but it turns out it's a bit nicer this way. We need ownership of parts of the slot (e.g. all buckets), with a rotation we then have to `std::mem::replace` the parts we need ownership of. We also don't get around a fallible access/unwrap like here, since we still need a bounds check.

Removing the item (which just shifts indices internally and returns the value) and adding a new one back turned out to be nicer.